README.md
# Cryptology

[![Gem Version](https://badge.fury.io/rb/cryptology.svg)](https://badge.fury.io/rb/cryptology)
[![Build Status](https://travis-ci.com/rubysamurai/cryptology.svg?branch=master)](https://travis-ci.com/rubysamurai/cryptology)

`Cryptology` is a wrapper for encryption and decryption in Ruby using OpenSSL. By default `AES-256-CBC` cipher is used.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'cryptology'
```

Save `Gemfile` and execute `$ bundle` command to install the gem.

Or to install it yourself run this command:

```
$ gem install cryptology
```

## Usage

```ruby
# Encrypting
Cryptology.encrypt(data: data, key: key, salt: salt, iter: iter, cipher: cipher, iv: iv)

# Decrypting
Cryptology.decrypt(data: data, key: key, salt: salt, iter: iter, cipher: cipher, iv: iv)

# Check decryption ability (true if can be decrypted, false otherwise)
Cryptology.decryptable?(data: data, key: key, salt: salt, iter: iter, cipher: cipher, iv: iv)
```


Argument | Required? | Default                 | Comment
---------|-----------|-------------------------|-------------
data     | **Yes**   | n/a                     | Data to encrypt or decrypt
key      | **Yes**   | n/a                     | Secure key for encryption and decryption
salt     | *No*      | Random 16 bytes         | Value to prevent attacks on key based on dictionaries
iter     | *No*      | 10,000                  | Number of iterations to adjust computation time
cipher   | *No*      | `AES-256-CBC`           | Cipher algorithm
iv       | *No*      | Random iv for algorithm | Initialization vector

Example:

```ruby
# Data to encrypt (required)
data = 'Very, very confidential data'

# Secure key for encryption (required)
key = 'password_01X'

# Salt (optional)
salt = OpenSSL::Random.random_bytes(16)
# => "r\x97\xEA9]I\x18\x05\xEAZ\xA2\xBB^Y=\x83"

# Number of iterations (optional)
iter = 50000

# Use Camellia cipher in CBC mode (optional)
cipher = 'CAMELLIA-256-CBC'

# Initialization vector for CAMELLIA-256-CBC (optional)
iv = OpenSSL::Cipher.new(cipher).random_iv
# => "\xB0\xCA\xBBc5'\x03i\x01\xC1@\xC0\xB6\xCE7+"

# Encrypt our data
enc = Cryptology.encrypt(data: data,
                         key: key,
                         salt: salt,
                         iter: iter,
                         cipher: cipher,
                         iv: iv)

# => { "cipher"=>"CAMELLIA-256-CBC",
#      "salt"=>"r\x97\xEA9]I\x18\x05\xEAZ\xA2\xBB^Y=\x83",
#      "iter"=>50000,
#      "iv"=>"\xB0\xCA\xBBc5'\x03i\x01\xC1@\xC0\xB6\xCE7+",
#      "data"=>"k+e3JZpkFIgkB15LjK85k5roojNgawN9yPEp6CXGhCQ=\n" }

# Verify that data can be decrypted
Cryptology.decryptable?(data: enc['data'],
                        key: key,
                        salt: enc['salt'],
                        iter: enc['iter'],
                        cipher: enc['cipher'],
                        iv: enc['iv'])
#  => true

# Decrypt our data
plain = Cryptology.decrypt(data: enc['data'],
                           key: key,
                           salt: enc['salt'],
                           iter: enc['iter'],
                           cipher: enc['cipher'],
                           iv: enc['iv'])
# => "Very, very confidential data"
```

### Cipher algorithms

> **Note:** Ruby 2.4 and above would throw an error if key is too short or too long for a given cipher algorithm (see [this commit](https://github.com/ruby/ruby/commit/ce635262f53b760284d56bb1027baebaaec175d1) for details) Make sure you choose a cipher with 32 bytes size key.

List of tested and supported ciphers:

```
Ruby 2.7.0, OpenSSL 1.1.1

AES-128-XTS
AES-256-CBC
AES-256-CBC-HMAC-SHA1
AES-256-CBC-HMAC-SHA256
AES-256-CFB
AES-256-CFB1
AES-256-CFB8
AES-256-CTR
AES-256-ECB
AES-256-OFB
AES256

ARIA-256-CBC
ARIA-256-CFB
ARIA-256-CFB1
ARIA-256-CFB8
ARIA-256-CTR
ARIA-256-ECB
ARIA-256-OFB
ARIA256

CAMELLIA-256-CBC
CAMELLIA-256-CFB
CAMELLIA-256-CFB1
CAMELLIA-256-CFB8
CAMELLIA-256-CTR
CAMELLIA-256-ECB
CAMELLIA-256-OFB
CAMELLIA256

CHACHA20
CHACHA20-POLY1305
```

## License

`Cryptology` © Dmitriy Tarasov. Released under the [MIT](LICENSE.txt) license.


CHANGELOG.md
# [2.1.0](https://github.com/rubysamurai/cryptology/compare/v2.0.0...v2.1.0) (2020-03-04)


### Features

* Require Ruby 2.4 ([796d1a1](https://github.com/rubysamurai/cryptology/commit/796d1a1ae6e7bcbc2a0062108f8aec0ed68a6d58))



# [2.0.0](https://github.com/rubysamurai/cryptology/compare/v1.2.0...v2.0.0) (2017-06-27)


### Features

* `encrypt` method returns a hash ([a392972](https://github.com/rubysamurai/cryptology/commit/a3929722c089c6871de9afb91d3199f35a883997))
* Always set `iv` when possible ([a392972](https://github.com/rubysamurai/cryptology/commit/a3929722c089c6871de9afb91d3199f35a883997))
* Digest `key` with PBKDF2 SHA256 with an optional `salt` argument ([2296d04](https://github.com/rubysamurai/cryptology/commit/2296d04cd4834d09d0cb1be8b40688b7ce25fa36))
* Optional `iter` argument ([a3a774d](https://github.com/rubysamurai/cryptology/commit/a3a774d87befa373d45274b350687017a8861abe))



# [1.2.0](https://github.com/rubysamurai/cryptology/compare/v1.1.0...v1.2.0) (2015-12-10)


### Features

* `decryptable?` method ([610dfce](https://github.com/rubysamurai/cryptology/commit/610dfce5f24238327dac85bbf0bca41ad8309dbf))



# [1.1.0](https://github.com/rubysamurai/cryptology/compare/v1.0.1...v1.1.0) (2015-12-08)


### Features

* Require Ruby 2.1 ([bce1367](https://github.com/rubysamurai/cryptology/commit/bce1367c51a40239de39c0c7464c6666356a75c1))
* Use keyword arguments ([31de476](https://github.com/rubysamurai/cryptology/commit/31de476d3057101efef95fb3378b27fdf9ad9e32))



# [1.0.1](https://github.com/rubysamurai/cryptology/compare/v1.0.0...v1.0.1) (2015-12-08)


### Bug Fixes

* Allow `iv` to be nil ([5abf3e1](https://github.com/rubysamurai/cryptology/commit/5abf3e1f5f40eebd0b38be6b2b57fac642c83cb4))



# 1.0.0 (2015-12-07)


Initial release.


spec/spec_helper.rb
require 'bundler/setup'
require 'cryptology'

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = '.rspec_status'

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/cryptology_spec.rb
CIPHERS = %w[AES-128-XTS
             AES-256-CBC
             AES-256-CBC-HMAC-SHA1
             AES-256-CBC-HMAC-SHA256
             AES-256-CFB
             AES-256-CFB1
             AES-256-CFB8
             AES-256-CTR
             AES-256-ECB
             AES-256-OFB
             AES256

             ARIA-256-CBC
             ARIA-256-CFB
             ARIA-256-CFB1
             ARIA-256-CFB8
             ARIA-256-CTR
             ARIA-256-ECB
             ARIA-256-OFB
             ARIA256

             CAMELLIA-256-CBC
             CAMELLIA-256-CFB
             CAMELLIA-256-CFB1
             CAMELLIA-256-CFB8
             CAMELLIA-256-CTR
             CAMELLIA-256-ECB
             CAMELLIA-256-OFB
             CAMELLIA256

             CHACHA20
             CHACHA20-POLY1305].freeze

RSpec.describe Cryptology do
  let(:data) { 'Very confidential data with UTF-8 symbols: ♠ я ü æ' }
  let(:key) { 'veryLongAndSecurePassword_61543054534' }
  let(:salt) { OpenSSL::Random.random_bytes(16) }
  let(:iter) { 25_000 }
  let(:cipher) { 'CAMELLIA-256-CBC' }
  let(:iv) { OpenSSL::Cipher.new(cipher).random_iv }

  it 'encrypts and decrypts with default arguments' do
    encrypted = Cryptology.encrypt(data: data, key: key)
    expect(
      Cryptology.decrypt(data: encrypted['data'],
                         key: key,
                         salt: encrypted['salt'],
                         iv: encrypted['iv'])
    ).to eq data
  end

  it 'encrypts and decrypts with salt argument' do
    encrypted = Cryptology.encrypt(data: data, key: key, salt: salt)
    expect(encrypted['salt']).to eq salt
    expect(
      Cryptology.decrypt(data: encrypted['data'],
                         key: key,
                         salt: salt,
                         iv: encrypted['iv'])
    ).to eq data
  end

  it 'encrypts and decrypts with iter argument' do
    encrypted = Cryptology.encrypt(data: data, key: key, iter: iter)
    expect(encrypted['iter']).to eq iter
    expect(
      Cryptology.decrypt(data: encrypted['data'],
                         key: key,
                         salt: encrypted['salt'],
                         iter: iter,
                         iv: encrypted['iv'])
    ).to eq data
  end

  it 'encrypts and decrypts with iv argument' do
    encrypted = Cryptology.encrypt(data: data, key: key, iv: iv)
    expect(encrypted['iv']).to eq iv
    expect(
      Cryptology.decrypt(data: encrypted['data'],
                         key: key,
                         salt: encrypted['salt'],
                         iv: iv)
    ).to eq data
  end

  it 'encrypts and decrypts with cipher argument' do
    encrypted = Cryptology.encrypt(data: data, key: key, cipher: cipher)
    expect(encrypted['cipher']).to eq cipher
    expect(
      Cryptology.decrypt(data: encrypted['data'],
                         key: key,
                         salt: encrypted['salt'],
                         cipher: cipher,
                         iv: encrypted['iv'])
    ).to eq data
  end

  context '#encrypt' do
    it 'throws error without data argument' do
      expect { Cryptology.encrypt(key: key) }
        .to raise_error(ArgumentError)
    end

    it 'throws error without key argument' do
      expect { Cryptology.encrypt(data: data) }
        .to raise_error(ArgumentError)
    end
  end

  context '#decrypt' do
    let(:encrypted) { Cryptology.encrypt(data: data, key: key) }

    it 'throws error without data argument' do
      expect { Cryptology.decrypt(key: key, salt: salt, iv: iv) }
        .to raise_error(ArgumentError)
    end

    it 'throws error without key argument' do
      expect { Cryptology.decrypt(data: encrypted, salt: salt, iv: iv) }
        .to raise_error(ArgumentError)
    end

    it 'throws error without salt argument' do
      expect { Cryptology.decrypt(data: encrypted, key: key, iv: iv) }
        .to raise_error(ArgumentError)
    end

    it 'throws error without iv argument' do
      expect { Cryptology.decrypt(data: encrypted, key: key, salt: salt) }
        .to raise_error(ArgumentError)
    end
  end

  context '#decryptable?' do
    let(:encrypted) do
      Cryptology.encrypt(data: data, key: key, salt: salt, iter: iter, cipher: cipher, iv: iv)
    end

    it 'returns true for valid arguments' do
      expect(
        Cryptology.decryptable?(data: encrypted['data'],
                                key: key,
                                salt: salt,
                                iter: iter,
                                cipher: cipher,
                                iv: iv)
      ).to be true
    end

    it 'returns false for invalid data argument' do
      expect(
        Cryptology.decryptable?(data: 'random',
                                key: key,
                                salt: salt,
                                iter: iter,
                                cipher: cipher,
                                iv: iv)
      ).to be false
    end

    it 'returns false for invalid key argument' do
      expect(
        Cryptology.decryptable?(data: encrypted['data'],
                                key: key.upcase,
                                salt: salt,
                                iter: iter,
                                cipher: cipher,
                                iv: iv)
      ).to be false
    end

    it 'returns false for invalid salt argument' do
      expect(
        Cryptology.decryptable?(data: encrypted['data'],
                                key: key,
                                salt: 'invalid',
                                iter: iter,
                                cipher: cipher,
                                iv: iv)
      ).to be false
    end

    it 'returns false for invalid iter argument' do
      expect(
        Cryptology.decryptable?(data: encrypted['data'],
                                key: key,
                                salt: salt,
                                iter: iter - 1,
                                cipher: cipher,
                                iv: iv)
      ).to be false
    end

    it 'returns false for invalid cipher argument' do
      expect(
        Cryptology.decryptable?(data: encrypted['data'],
                                key: key,
                                salt: salt,
                                iter: iter,
                                cipher: 'AES-256-CBC',
                                iv: iv)
      ).to be false
    end
  end

  context 'encryption' do
    CIPHERS.each do |c|
      it "encrypts #{c}" do
        data = 'Very, very confidential data'
        key  = OpenSSL::Cipher.new(c).random_key
        expect { Cryptology.encrypt(data: data, key: key, cipher: c) }.not_to raise_error
      end
    end
  end

  context 'decryption' do
    CIPHERS.each do |c|
      it "decrypts #{c}" do
        data      = 'Very confidential data with UTF-8 symbols: ♠ я ü æ'
        key       = OpenSSL::Cipher.new(c).random_key
        encrypted = Cryptology.encrypt(data: data, key: key, cipher: c)
        expect(
          Cryptology.decrypt(data: encrypted['data'],
                             key: key,
                             salt: encrypted['salt'],
                             iter: encrypted['iter'],
                             cipher: c,
                             iv: encrypted['iv'])
        ).to eq data
      end
    end
  end
end


lib/cryptology.rb
require 'cryptology/version'
require 'openssl'
require 'base64'

module Cryptology
  def self.encrypt(data:, key:, salt: nil, iter: 10_000, cipher: 'AES-256-CBC', iv: nil)
    salt ||= OpenSSL::Random.random_bytes(16)
    iv ||= OpenSSL::Cipher.new(cipher).random_iv
    encrypted = encrypt_data(data.to_s, digest_key(key, salt, iter), cipher, iv)
    { 'cipher' => cipher,
      'salt' => salt,
      'iter' => iter,
      'iv' => iv,
      'data' => ::Base64.encode64(encrypted) }
  end

  def self.decrypt(data:, key:, salt:, iter: 10_000, cipher: 'AES-256-CBC', iv:)
    base64_decoded = ::Base64.decode64(data.to_s)
    decrypt_data(base64_decoded, digest_key(key, salt, iter), cipher, iv)
      .force_encoding('UTF-8').encode
  end

  def self.decryptable?(data:, key:, salt:, iter: 10_000, cipher: 'AES-256-CBC', iv:)
    return true if decrypt(data: data, key: key, salt: salt, iter: iter, cipher: cipher, iv: iv)
  rescue OpenSSL::Cipher::CipherError
    false
  end

  def self.encrypt_data(data, key, cipher, iv)
    c = OpenSSL::Cipher.new(cipher)
    c.encrypt
    c.key = key
    c.iv  = iv unless iv.length != c.random_iv.length
    c.update(data) + c.final
  end

  def self.decrypt_data(data, key, cipher, iv)
    decipher = OpenSSL::Cipher.new(cipher)
    decipher.decrypt
    decipher.key = key
    decipher.iv  = iv unless iv.length != decipher.random_iv.length
    decipher.update(data) + decipher.final
  end

  def self.digest_key(key, salt, iter)
    digest = OpenSSL::Digest::SHA256.new
    len = digest.digest_length
    OpenSSL::PKCS5.pbkdf2_hmac(key, salt, iter, len, digest)
  end

  private_class_method :encrypt_data, :decrypt_data, :digest_key
end
